jax/_src/lax/lax.py

```diff
@@ -4325,11 +4341,49 @@ def _iota_abstract_eval(*, dtype, shape, dimension):
 iota_p.def_impl(partial(xla.apply_primitive, iota_p))
 iota_p.def_abstract_eval(_iota_abstract_eval)
 
-def _iota_lower(ctx, *, dtype, shape, dimension):
-  del dtype, shape
+def _iota_staging_rule(trace, *dyn_shape, dtype, shape, dimension):
+  params = dict(dtype=dtype, shape=shape, dimension=dimension)
+  if not dyn_shape:
+    return trace.default_process_primitive(iota_p, (), params)
+  source_info = source_info_util.current()
+  ds = iter(dyn_shape)
+  out_shape_for_tracer = [next(ds) if d is None else d for d in shape]
+  aval = core.DShapedArray(tuple(out_shape_for_tracer), dtype, False)
+  out_tracer = pe.DynamicJaxprTracer(trace, aval, source_info)
+  invars = [trace.getvar(d) for d in dyn_shape]
+  eqn = pe.new_jaxpr_eqn(invars, [trace.makevar(out_tracer)],
+                         iota_p, params, core.no_effects, source_info)
+  trace.frame.eqns.append(eqn)
```

Comment:
should use `frame.add_eqn` (there are no effects here but this method is more future-proof)

Reply:
Good idea, done.